binary-extensions: update Windows instructions (#911)
pypa/packaging.python.org: source/guides/packaging-binary-extensions.rst

Before it is possible to build a binary extension, it is necessary to ensure
that you have a suitable compiler available. On Windows, Visual C is used to
build the official CPython interpreter, and should be used to build compatible
binary extensions.  To set up a build environment for binary extensions, install
`Visual Studio Community Edition <https://www.visualstudio.com/en-us/downloads/download-visual-studio-vs.aspx>`__
- any recent version is fine.

One caveat: if you use Visual Studio 2019 or later, your extension will depend
on an "extra" file, ``VCRUNTIME140_1.dll``, in addition to the
``VCRUNTIME140.dll`` that all previous versions back to 2015 depend on. This
will add an extra requirement to using your extension on versions of CPython
that do not include this extra file. To avoid this, you can add the

> compile-time argument ``/d2FH4-``. Recent versions of Python may include this

Author: I'd like to put exact numbers, but I'm not sure which ones they are (it might have changed in patch releases, too?)

build the official CPython interpreter, and should be used to build compatible
binary extensions.

> Python 2.7 used Visual Studio 2008, and Python 3.5+ uses Visual Studio 2015 or
> later; since 2015 Microsoft has kept backward compatibility (with one exception,
> listed below). Unfortunately, older versions of Visual Studio are no longer
> available from Microsoft, so building for CPython < 3.5 is discouraged. If you do

Reviewer: I'm not sure the extra detail is necessary here (though maybe a mention of Visual Studio 2015 would still be useful).

```suggestion
Building for Python prior to 3.5 is discouraged, because older versions of
Visual Studio are no longer available from Microsoft. If you do
```

need to build for older versions, you can set ``DISTUTILS_USE_SDK=1`` and
``MSSdk=1`` to force a the currently activated version of MSVC to be found,

> and you should exercize care when designing your extension not to malloc/free

Reviewer: ```suggestion
and you should exercise care when designing your extension not to malloc/free
```

memory across different libraries, avoid relying on changed data structures,
and so on. Tools for generating extension modules usually avoid these things
for you.

To set up a build environment for binary extensions, install
`Visual Studio Community Edition <https://www.visualstudio.com/en-us/downloads/download-visual-studio-vs.aspx>`__ - any

> recent version is fine.

Reviewer: This feels like the most important bit. Could it go before the "older versions" paragraph above?

Author: Moved.

> One caveat; if you use MSVC 2019+, your extension will depend on an "extra"
> file, ``VCRUNTIME140_1.dll``, in addition to the ``VCRUNTIME140.dll`` all previous

Reviewer: ```suggestion
One caveat: if you use Visual Studio 2019 or later, your extension will depend on an "extra"
file, ``VCRUNTIME140_1.dll``, in addition to the ``VCRUNTIME140.dll`` that all previous
```

verisions back to 2015 depend on. This will add an extra requirement to using

> your extension on versions of CPython that do not include this extra file; you
> can add the compile-time argument ``/d2FH4-`` to avoid this link on 2019+.

Reviewer: ```suggestion
your extension on versions of CPython that do not include this extra file. To
avoid this, you can add the compile-time argument ``/d2FH4-``.
```

> Building with the recommended compiler on Windows ensures that a compatible C library
> is used throughout the Python process.

Reviewer: This feel unnecessary, or it could be conveyed introductory paragraph.

```suggestion
```
